src/contexts/services/User.tsx

```diff
@@ -1,68 +1,141 @@
-import React, { createContext, useContext, useEffect, useMemo, useReducer } from 'react';
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
 
-import { FetchedUser, User } from '@app/types';
+import { User, UserCreationData } from '@app/types';
 
-import UserReducer, { UserAction } from '@app/reducers/User.Reducer';
-import useFirebaseAuth from '@app/hooks/useFirebaseAuth';
-import useUserCollection from '@app/hooks/useUserCollection';
-import { logUser, usersCollection } from '@app/firebase/firebase';
+import { auth, createUserWithData, store } from '@app/firebase/firebase';
 
-import { loggedOutUser } from './utils/UserProps';
+interface PromiseNotification {
+  resolve?: (result?: AuthResult) => void;
+  reject?: (err: AuthError) => void;
+}
+
+class AuthError extends Error {
+  constructor(public readonly code: string | number, public readonly message: string) {
+    super(message);
+    this.name = 'R51AuthError';
+  }
+}
+
+type AuthResult = { status?: 'SUCCESS' | 'PARTIAL_SUCCESS' } | undefined | void;
 
-export interface UserCtx {
-  user: User;
-  refreshToken: () => Promise<string>;
-  userDispatch: React.Dispatch<UserAction>;
-  isLoggingIn: boolean;
-  setIsLoggingIn: (next: boolean) => void;
-  usersRequestingVerify: FetchedUser[];
-  verifiedResidents: FetchedUser[];
+interface UserCtx {
+  user: User | null | undefined;
+  signIn: (email: string, password: string) => Promise<AuthResult>;
+  signOut: () => Promise<AuthResult>;
+  signUp: (userData: UserCreationData) => Promise<AuthResult>;
 }
 
-export const UserContext = createContext({} as UserCtx);
-export const useUser = (): UserCtx => useContext<UserCtx>(UserContext);
+const UserContext = createContext({} as UserCtx);
+export const useUser = (): UserCtx => useContext(UserContext);
+
+const UserContextProvider: React.FC = ({ children }) => {
+  const [user, setUser] = useState<User | null | undefined>();
```

Comment:
I meant to discuss this on the last call but during development of 0.1 I realized it was more convenient to still `user` to an object even while logged out. If I could make my case:

1. Even anonymous users can be given metadata that is useful to the application.
2. I was constantly forgetting whether `null` or `undefined` was the logging- in or the unauthenticated user value
3. It prevents the need for as much optional chaining during user checks (by personal preference I find it kind of ugly), and
4. It will allow for cleaner comparisons when determining access on the frontend - most commonly I found myself using things like (if user.permissions < 2) instead of having to use null checks or optional chaining.

Reply:
Makes sense 16fe7527d4f82787c91b1d13173a36622d772426.

Reply:
This doesn't have anything to do with the suggestion, just a side: I really like optional chaining syntax. When reading through a?.b?.c you can kind of go "Is a there -> give me b -> is b there -> give me c". I think it reads really well.

Reply:
Oh don't get me wrong, there are plenty of cases I agree optional chaining is a lifesaver - I'll be the last guy to advocate for the days of:
```javascript
myInexplicablyLongPossiblyNullVariable && myInexplicablyLongPossiblyNullVariable.property
```
But I've found in some cases it makes it too tempting to make a value possibly null when it doesn't need to be